8356218: [macos] Document --app-content

sashamatveev opened the pull request:
- Added following note for `--app-content` on macOS to help and man page: `The value should be a directory with the "Resources" subdirectory (or any other directory that is valid in the "Contents" directory of the application bundle). Otherwise, jpackage may produce invalid application bundle which may fail code signing and/or notarization.`
- Added warning if `--app-content` if it points to non-standard subdirectory in "Contents" directory.
- Added test to cover warning message.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8356218](https://bugs.openjdk.org/browse/JDK-8356218): [macos] Document --app-content (**Enhancement** - P3)


### Reviewers
 * [Alexey Semenyuk](https://openjdk.org/census#asemenyuk) (@alexeysemenyukoracle - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/26848/head:pull/26848` \
`$ git checkout pull/26848`

Update a local copy of the PR: \
`$ git checkout pull/26848` \
`$ git pull https://git.openjdk.org/jdk.git pull/26848/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 26848`

View PR using the GUI difftool: \
`$ git pr show -t 26848`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/26848.diff">https://git.openjdk.org/jdk/pull/26848.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/26848#issuecomment-3202083404)
</details>

alexeysemenyukoracle commented on src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacApplicationBuilder.java:
```diff
@@ -123,6 +126,18 @@ private static void validateAppVersion(Application app) throws ConfigException {
+            if (!CONTENTS_SUB_DIRS.stream()
+                    .anyMatch(subDir -> contentDir.getFileName().toString()
+                                                  .equalsIgnoreCase(subDir))) {
+                Log.info(MessageFormat.format(I18N.getString(
```
Can be simplified down to:
```
Log.info(I18N.format("warning.non.standard.contents.sub.dir", contentDir));
```

sashamatveev replied:
Fixed. I copy-paste this code. We have many places with `MessageFormat.format(I18N.getString())`. Should I file follow up bug to clean-up it?

sashamatveev commented on src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacApplicationBuilder.java:
```diff
@@ -123,6 +126,18 @@ private static void validateAppVersion(Application app) throws ConfigException {
+        for (var contentDir : app.contentDirs()) {
+            if (!CONTENTS_SUB_DIRS.stream()
+                    .anyMatch(subDir -> contentDir.getFileName().toString()
+                                                  .equalsIgnoreCase(subDir))) {
```
By default macOS is case-insensitive. If I rename `MacOS` to `macos` in application bundle, such bundle will still works. I assumed that `Resources` and `resources` are same thing. After thinking more about it I think it should be case sensitive comparison. Will change it.

sashamatveev replied:
Changed to case sensitive comparison.

fandreuz commented on src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/MacApplicationBuilder.java:
```diff
@@ -123,6 +125,17 @@ private static void validateAppVersion(Application app) throws ConfigException {
 
+    private static void validateAppContentDirs(Application app) {
+        for (var contentDir : app.contentDirs()) {
+            if (!CONTENTS_SUB_DIRS.stream()
```
You could replace `CONTENTS_SUB_DIRS` with a `Set<String>`, and the inspection with `CONTENTS_SUB_DIRS.contains(contentDir.getFileName().toString())`.

sashamatveev replied:
Updated as suggested.

sashamatveev commented on src/jdk.jpackage/macosx/classes/jdk/jpackage/internal/resources/MacResources.properties:
```diff
@@ -85,3 +85,4 @@ message.codesign.failed.reason.app.content="codesign" failed and additional appl
 message.codesign.failed.reason.xcode.tools=Possible reason for "codesign" failure is missing Xcode with command line developer tools. Install Xcode with command line developer tools to see if it resolves the problem.
 warning.unsigned.app.image=Warning: Using unsigned app-image to build signed {0}.
 warning.per.user.app.image.signed=Warning: Support for per-user configuration of the installed application will not be supported due to missing "{0}" in predefined signed application image.
+warning.non.standard.contents.sub.dir=Warning: --app-content value "{0}" points to the non-standard subdirectory in the "Contents" directory of the application bundle.
```
Updated as suggested.

sashamatveev commented:
8356218: [macos] Document --app-content [v3]
- Fixed latest review comments.
- Test moved from `ErrorTest` to `AppContentTest`.
- Added `setIgnoreExitCode()` to `JPackageCommand` to ignore exit code. Added test case may or may not fail depending on macOS version.

alexeysemenyukoracle commented on test/jdk/tools/jpackage/helpers/jdk/jpackage/test/JPackageCommand.java:
```diff
@@ -742,6 +742,11 @@ public JPackageCommand validateOutput(Consumer<Iterator<String>> validator) {
         return this;
     }
 
+    public JPackageCommand setIgnoreExitCode(boolean v) {
+        ignoreExitCode = v;
+        return this;
+    }
```
This is bad. It makes unclear semantics of the `JPackageCommand.execute(int expectedExitCode)` function.

I suggest adding `JPackageCommand.executeIgnoreExitCode()` function instead in the following way:
```
public Executor.Result executeIgnoreExitCode() {
    return execute(Optional.empty());
}

public Executor.Result execute(int expectedExitCode) {
    return execute(Optional.of(expectedExitCode));
} 

private Executor.Result execute(Optional expectedExitCode) {
    // Here goes the implementation of the current execute(int expectedExitCode) function adjusted accordingly
} 
```

sashamatveev replied:
Fixed.
